lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,773 @@
+		let input = TxIn {
+			previous_output: OutPoint {
+				txid: tx.txid(),
+				vout: msg.prevtx_out,
+			},
+			sequence: Sequence(msg.sequence),
+			..Default::default()
+		};
+		debug_assert!((msg.prevtx_out as usize) < tx.output.len());
+		let prev_output = &tx.output[msg.prevtx_out as usize];
+		self.inputs.insert(msg.serial_id, TxInputWithPrevOutput {
+			input,
+			prev_output: prev_output.clone(),
+		});
+	}
+
+	fn local_tx_add_output(&mut self, msg: &msgs::TxAddOutput) {
+		self.outputs.insert(msg.serial_id, TxOut {
+			value: msg.sats,
+			script_pubkey: msg.script.clone(),
+		});
+	}
+
+	fn local_tx_remove_input(&mut self, msg: &msgs::TxRemoveInput) {
+		self.inputs.remove(&msg.serial_id);
+	}
+
+	fn local_tx_remove_output(&mut self, msg: &msgs::TxRemoveOutput) {
+		self.outputs.remove(&msg.serial_id);
+	}
+
+	fn build_transaction(self) -> Result<Transaction, AbortReason> {
+		// The receiving node:
+		// MUST fail the negotiation if:
+
+		// - the peer's total input satoshis is less than their outputs
+		let counterparty_inputs_contributed = self.counterparty_inputs_contributed();
+		let counterparty_inputs_value: u64 = counterparty_inputs_contributed.clone()
+			.map(|input| input.prev_output.value).sum();
+		let counterparty_outputs_contributed = self.counterparty_outputs_contributed();
+		let counterparty_outputs_value: u64 = counterparty_outputs_contributed.clone()
+			.map(|output| output.value).sum();
+		if counterparty_inputs_value < counterparty_outputs_value {
+			return Err(AbortReason::OutputsExceedInputs);
+		}
+
+		// - there are more than 252 inputs
+		// - there are more than 252 outputs
+		if self.inputs.len() > MAX_INPUTS_OUTPUTS_COUNT || self.outputs.len() > MAX_INPUTS_OUTPUTS_COUNT {
+			return Err(AbortReason::ExceededNumberOfInputsOrOutputs);
+		}
+
+		let tx_to_validate = Transaction {
+			version: 2,
+			lock_time: self.tx_locktime,
+			input: self.inputs.values().map(|p| p.input.clone()).collect(),
+			output: self.outputs.values().cloned().collect(),
+		};
+		if tx_to_validate.weight() as u32 > MAX_STANDARD_TX_WEIGHT {
+			return Err(AbortReason::TransactionTooLarge);
+		}
+
+		// TODO: How do we enforce their fees cover the witness without knowing its expected length?
+		const INPUT_WEIGHT: u64 = BASE_INPUT_WEIGHT + EMPTY_SCRIPT_SIG_WEIGHT;
+
+		// - the peer's paid feerate does not meet or exceed the agreed feerate (based on the minimum fee).
+		let counterparty_output_weight_contributed: u64 = counterparty_outputs_contributed.clone().map(|output|
+			(8 /* value */ + output.script_pubkey.consensus_encode(&mut sink()).unwrap() as u64) *
+				WITNESS_SCALE_FACTOR as u64
+		).sum();
+		let counterparty_weight_contributed = counterparty_output_weight_contributed +
+			counterparty_inputs_contributed.clone().count() as u64 * INPUT_WEIGHT;
+		let counterparty_fees_contributed =
+			counterparty_inputs_value.saturating_sub(counterparty_outputs_value);
+		let mut required_counterparty_contribution_fee =
```

Comment:
This can lead to division by zero, special case must be handles earlier, precondition

Reply:
Actually I believe the calculation should be `self.feerate_sat_per_kw as u64 * counterparty_weight_contributed / 1000`.

`sat/kw * w * 1kw/1000w = sat`

So the numerator and denominator just need to swap